crates/bitcoin/src/formatter.rs

```diff
@@ -6,234 +6,292 @@ use crate::{merkle::MerkleProof, script::*, types::*, Error, GetCompact};
 const WITNESS_FLAG: u8 = 0x01;
 const WITNESS_MARKER: u8 = 0x00;
 
-/// Type to be formatted as a bytes array
-pub trait Formattable<Options = ()> {
-    fn format(&self) -> Vec<u8>;
-    fn format_with(&self, _options: Options) -> Vec<u8> {
-        self.format()
+trait CheckedReduce: Sized {
+    fn checked_reduce(&mut self, amount: Self) -> Result<(), Error>;
+}
+
+impl CheckedReduce for u32 {
+    fn checked_reduce(&mut self, amount: Self) -> Result<(), Error> {
+        self.checked_sub(amount)
+            .map(|new_self| *self = new_self)
+            .ok_or(Error::ArithmeticUnderflow)
     }
 }
 
-pub trait TryFormattable<Options = ()> {
-    fn try_format(&self) -> Result<Vec<u8>, Error>;
-    fn try_format_with(&self, _options: Options) -> Result<Vec<u8>, Error> {
-        self.try_format()
+/// Type to be formatted as a bytes array
+pub trait TryFormat<Options = ()>: Sized {
+    fn format_size(&self) -> u32 {
+        sp_std::mem::size_of::<Self>() as u32
+    }
```

Comment:
I'm not sure I like having default implementation, it makes it easy to forget to properly implement for the various types

Reply:
Done, but it's not the most elegant solution. Let me know what you think